### PR TITLE
Update to support casper 2.0.0-rc4

### PR DESCRIPTION
**Pull request by Avi-D-coder**
Deletes a bunch of stuff.
Makes a new version.

Other than using upstream types the main breaking change is replacing `casper_node::JsonBlock` with `casper_types::JsonBlockWithSignatures`.

`JsonBlock` has been removed from `casper_node`, it was briefly added to `casper_sidecar`, but then it was removed. I believe everything now uses `JsonBlockWithSignatures`. 
`JsonBlockWithSignatures` sounds like it should be the same type when serialized, but it's rep is very different.

Upstream users of this library are unlikely to be able to continue getting `JsonBlock`s from the l1, and should now update to use `JsonBlockWithSignatures`.

This change has been made in litmus-wasm, but the service worker will also have to update.

I did attempt to support an updated V2 compatible version of `JsonBlock`, but it's a pain, and is still a breaking change so I stopped, it's best if that change is done in the JS.

### CODE REVIEW

File: src/block.rs [hunk 1]
```diff
@@ -6,18 +6,17 @@ use casper_types::{crypto::gens::public_key_arb, SecretKey};
...
 use casper_types::{
-    bytesrepr::{FromBytes, ToBytes},
-    EraId, PublicKey, Signature,
+    bytesrepr::{self, FromBytes, ToBytes},
+    EraId, PublicKey, RewardedSignatures, Signature, TransactionHash,
 };
 
 use super::{
-    block_header::{BlockHash, BlockHeader},
+    block_header::{BlockHash, BlockHeaderV1},
```
Comment (xcthulhu) [inline src/block.rs:14]: We can just trash `BlockHeaderV1` and bump the semvar in `Cargo.toml`

File: src/consensus.rs [hunk 2]
```diff
@@ -103,17 +112,21 @@ impl FromBytes for EraReport {
 #[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
 #[cfg_attr(test, derive(serde::Serialize, serde::Deserialize))]
 // See: https://github.com/casper-network/casper-node/blob/8ca9001dabba0dae95f92ad8c54eddd163200b5d/node/src/types/block.rs#L748-L753
-pub struct EraEnd {
+pub struct EraEndV1 {
```
Comment (xcthulhu) [inline src/consensus.rs:115]: Why V1?

Comment (xcthulhu) [reply src/consensus.rs:115]: We don't support V1 anywhere else, so I think we can drop support here and just do V2.

**Comment by xcthulhu** [comment]
We could just get rid of [src/crypto.rs](https://github.com/cspr-rad/litmus/pull/9/files#diff-35315b0daa0b6640dfe8f3f8c497b05f4e3aac96998ed13f79a6a600cfac4499) and use `casper-types` implementation.  However, this might make proptesting elsewhere harder...

File: tests/integration.rs [hunk 3]
```diff
@@ -1,135 +1,4 @@
...
-    EraId,
-};
-
-use once_cell::sync::Lazy;
-
-static BLOCKS_MAP: Lazy<BTreeMap<u64, JsonBlock>> = Lazy::new(|| {
-    let mut blocks = BTreeMap::new();
-    let cwd = std::env::current_dir().unwrap();
-    let this_files_directory = std::path::Path::new(&(file!())).parent().unwrap();
-    let blocks_path = cwd.join(this_files_directory).join("assets/blocks");
```
Comment (xcthulhu) [inline tests/integration.rs:21]: We can delete all of `assets/blocks`

**Comment by Avi-D-coder** [comment]
@xcthulhu All deleted

**Review by xcthulhu — APPROVED** [review]
LGTM

File: src/merkle_proof.rs [hunk 4]
```diff
@@ -173,6 +181,7 @@ impl FromBytes for PointerBlock {
     }
 }
 
+// FIXME not sure if this is used anywhere?
```
Comment (xcthulhu) [inline src/merkle_proof.rs:184]: We can delete `Trie` and `PointerBlock`, since we are importing `compute_state_hash` now.

Comment (Avi-D-coder) [reply src/merkle_proof.rs]: fixed

File: src/kernel.rs [hunk 5]
```diff
@@ -22,24 +14,23 @@ pub enum BlockSignaturesValidationError {
...
-impl From<SignatureVerificationError> for BlockSignaturesValidationError {
-    fn from(signature_verification_error: SignatureVerificationError) -> Self {
-        BlockSignaturesValidationError::SignatureVerificationError(signature_verification_error)
-    }
-}
+// impl From<SignatureVerificationError> for BlockSignaturesValidationError {
+//     fn from(signature_verification_error: SignatureVerificationError) -> Self {
+//         BlockSignaturesValidationError::SignatureVerificationError(signature_verification_error)
+//     }
+// }
```
Comment (xcthulhu) [inline src/kernel.rs:28]: We can delete this I think.

File: src/kernel.rs [hunk 6]
```diff
@@ -52,34 +43,31 @@ impl EraInfo {
         self.era_id
     }
 
+    #[allow(clippy::result_large_err)]
     pub fn validate(
         &self,
-        block_header_with_signatures: &BlockHeaderWithSignatures,
+        JsonBlockWithSignatures { block, proofs }: JsonBlockWithSignatures,
```
Comment (xcthulhu) [inline src/kernel.rs:49]: @uchu Does this work with what the new side-care outputs when you try to get block by height?

File: src/kernel.rs [hunk 7]
```diff
@@ -52,34 +43,31 @@ impl EraInfo {
...
         }
         // See: https://github.com/casper-network/casper-node/blob/8ca9001dabba0dae95f92ad8c54eddd163200b5d/node/src/types/block.rs#L2465-L2474
-        let mut signature_data = block_header.block_hash().as_ref().to_vec();
-        signature_data.extend_from_slice(&block_header.era_id().to_le_bytes());
-        for (public_key, signature) in block_header_with_signatures
-            .block_signatures()
-            .proofs()
-            .iter()
-        {
+        let mut signature_data = block.hash().inner().into_vec();
```
Comment (xcthulhu) [inline src/kernel.rs:59]: So the `hash` on a block could possibly be forged (this accessor doesn't compute the hash, it access it from the `Block` data structure).

You need to add a check that `block.block_header().block_hash() == block.hash()`

Comment (xcthulhu) [reply src/kernel.rs:59]: You would also want to check the hash of the block body to make sure it matches the body hash in the header.

Comment (Avi-D-coder) [reply src/kernel.rs]: Fixed

File: src/kernel.rs (hunk 5 above)
Comment (Avi-D-coder) [reply src/kernel.rs]: I deleted this in the last commit.

**Pull request merged**